updated Router to properly translate uri dashes that map to controller subdirs

sneakyimp opened the pull request:
Fixes #4294 

**Description**
I have made a minor tweak to CodeIgniter\Router::validateRequest so that dashes in a subdirectory are mapped to underscores if `translateURIDashes`= `true`. This appears to correctly auto-route requests when you have a dash in the URI that corresponds to a controller directory.

**Checklist:**
- [x] Securely signed commits
- [x] Component(s) with PHPdocs
- [x] Unit testing, with >80% coverage
- [x] User guide updated
- [x] Conforms to style guide

sneakyimp commented:
@paulbalandan @kenjis @michalsn Could someone review this PR to fix #4294?

paulbalandan reviewed (commented):
I've looked at the changes and it looks good to me. I agree with the changes you've made. I have a minor cosmetic request. Change `$segmentConvert` to simply `$segment`. :)

I'll check first if this method is covered by tests or not before approving this for merge.

paulbalandan commented:
Can you add a test case for this one similar to here https://github.com/codeigniter4/CodeIgniter4/blob/develop/tests/system/Router/RouterTest.php#L215-L227 but using the underscored subfolder as in your case. It should assert that it will still get the correct controller and method.

sneakyimp commented:
Thank you, @paulbalandan!

> I've looked at the changes and it looks good to me. I agree with the changes you've made. I have a minor cosmetic request. Change `$segmentConvert` to simply `$segment`. :)

Please note that I chose a different variable name to distinguish this value from the raw segments referred to elsewhere. It was my feeling that this more clearly indicated that the segment had been changed with ucfirst and with the dash-to-underscore replacement. I'd like to keep it because I think it makes the intent of the code much clearer. It was pretty difficult to understand the intent of this code because there are so few comments and the variables and functions are poorly named. In particular, the function `validateRequest` is very poorly named. I'd very much like to give that function a different name but have not done so in order to expedite matters. Please let me know if I may keep this other var name for greater clarity?

I have another question: may I remove the ucfirst call in [setDirectory](https://github.com/codeigniter4/CodeIgniter4/blob/develop/system/Router/Router.php#L620)? This function is only called in the validateRequest function, where I have already applied ucfirst to the supplied parameter, and in [system/Test/FeatureTestTrait.php](https://github.com/codeigniter4/CodeIgniter4/blob/develop/system/Test/FeatureTestTrait.php#L191) where the supplied parameter is `null`.


> 
> I'll check first if this method is covered by tests or not before approving this for merge.

I have run the phpunit test and it passed all the previous tests, but please do run any additional tests:

`vendor/phpunit/phpunit/phpunit tests/system/Router`

I'll work on adding another test as you directed.

paulbalandan commented:
The tests run pretty smoothly and in fact the coverage increased for `Router` so that's good. But since we're fixing a bug here we'll want to document it that the fix has been well tested.

sneakyimp commented:
@paulbalandan I have committed another set of changes. NOTE that it implements more stringent enforcement of PSR-4 compliance in that it will throw PageNotFoundExceptions for controller names which are not acceptable PHP class names. This has the benefit of rejecting any `..` or `.` attempts which may sneak into the Router but it will also reject controller names with dashes or periods or puncutation, etc. This should make the Router a bit more efficient because it eliminates a fair amount of additional logic that was getting executed.

Note that I did change `RouterTest::testZeroAsURIPath()` to expect an Exception instead of passing. A zero is not a valid Controller name -- you can't declare a `class 0 {};` in PHP nor can you use a zero in any namespace declaration.

kenjis commented on system/Router/Router.php:
```diff
@@ -614,29 +632,53 @@ protected function validateRequest(array $segments): array
 		{
 			$this->directory = null;
-			return;
+			return true;
```
Why returns `true`?

sneakyimp replied:
Thank you so much for reviewing the code closely!

> Why returns `true`?

Oh no that's a mistake. I had considered having setDirectory return a boolean value if the directory was valid. I changed my mind but accidentally left this in there. I will remove the return value of true.

kenjis commented on system/Router/Router.php:
```diff
@@ -614,29 +632,53 @@ protected function validateRequest(array $segments): array
+	 */
+	private function isValidSegment(string $segment): bool
+	{
+		return (boolean)preg_match('/^[a-zA-Z_\x80-\xff][a-zA-Z0-9_\x80-\xff]*$/', $segment);
```
`(boolean)` is not used.

```suggestion
		return (bool) preg_match('/^[a-zA-Z_\x80-\xff][a-zA-Z0-9_\x80-\xff]*$/', $segment);
```

paulbalandan replied:
Or can be `preg_match('/^[a-zA-Z_\x80-\xff][a-zA-Z0-9_\x80-\xff]*$/', $segment) === 1`.

sneakyimp replied:
I'll change the cast to (bool).

paulbalandan commented on system/Router/Router.php:
```diff
@@ -573,32 +578,45 @@ public function autoRoute(string $uri)
+	 * @return array returns an array of remaining uri segments that don't map onto a directory
 	 */
-	protected function validateRequest(array $segments): array
+	protected function scanControllers(array $segments): array
```
As this method is not flagged as `@internal`, changing the name is technically BC breaking.

sneakyimp replied:
> As this method is not flagged as `@internal`, changing the name is technically BC breaking.

In order for this change to break any code, someone would have to have extended the router class. I did a grep search of the source and could find no place where it extends this class. This returns no results.
```
grep -r "extends.*Router" CodeIgniter4
```
If we are really concerned about it being a breaking change, perhaps we can define `validateRequest()` to simply call `scanControllers` and deprecate it:
```php
/**
 * @deprecated
 */
protected function validateRequest(array $segments): array
{
    return $this->scanControllers($segments);
}
```

If you want me to change the function name back to `validateRequest` I will do it, but it is a really bad name that does NOT describe what that function is doing.

paulbalandan replied:
But users may be extending the Router class for their own use case and we cannot be sure that they did not use this method. I'll go with the deprecation instead and reroute calls to scanController.

paulbalandan commented on system/Router/Router.php:
```diff
@@ -614,29 +632,53 @@ protected function validateRequest(array $segments): array
+			{
+				if (! $this->isValidSegment($segment))
+				{
+					return;
```
What is the real return type of this method? This line returns `void`, but line 644 returns `bool`.

sneakyimp replied:
`setDirectory` should not return anything. The return value of `true` is a mistake I made and I will remove it.

paulbalandan commented on system/Router/Router.php:
```diff
@@ -573,32 +578,45 @@ public function autoRoute(string $uri)
 
-		$c                 = count($segments);
-		$directoryOverride = isset($this->directory);
+		// if a prior directory value has been set, just return segments and get out of here
+		if (isset($this->directory))
+		{
+			return $segments;
+		}
```
Why is this necessary? As this is untested, can you add a test case covering this possibility.

sneakyimp replied:
I'm not entirely sure and will dig deeper. I definitely think we could use some more test cases and will try to add them.

paulbalandan commented on system/Router/Router.php:
```diff
@@ -614,29 +632,53 @@ protected function validateRequest(array $segments): array
 		}
 
-		$dir = ucfirst($dir);
+		if ($validate)
+		{
+			$segments = explode('/', trim($dir, '/'));
+			foreach ($segments as $segment)
+			{
+				if (! $this->isValidSegment($segment))
+				{
+					return;
+				}
+			}
+		}
```
This addition is not tested.

sneakyimp commented:
@paulbalandan I've pushed another commit and I think it resolves the issues you mentioned in your review -- for some reason github only offers me a "resolve conversation" button rather than continuing our discussion. I hope you will please take a look and let me know if you have further concerns.

paulbalandan commented on system/Router/Router.php:
```diff
@@ -614,29 +648,53 @@ protected function validateRequest(array $segments): array
+	 */
+	private function isValidSegment(string $segment): bool
+	{
+		return (bool)preg_match('/^[a-zA-Z_\x80-\xff][a-zA-Z0-9_\x80-\xff]*$/', $segment);
```
CS: Add space after the cast. `(bool) preg_match`

paulbalandan commented on system/Router/Router.php:
```diff
@@ -573,32 +578,61 @@ public function autoRoute(string $uri)
+			if (! is_file($test . '.php') && is_dir($test))
 			{
-				$this->setDirectory(array_shift($segments), true);
+				$this->setDirectory($segmentConvert, true, false);
```
I'm wondering. You added a third param to `setDirectory` optionally set to `true` but purposely set it to `false` here. As you have promoted a stricter check on the segments, no validation here seems a bit off. Or am I missing something?

sneakyimp commented:
@paulbalandan sorry I'm slow to respond. For some reason github isn't letting me respond directly under your review comment so I'm commenting here.

>CS: Add space after the cast. (bool) preg_match

I have added the space, committed to my local repo, rebased on the latest develop branch of CI4, and pushed. Please let me know if you have additional requests, questions, or issues.

>I'm wondering. You added a third param to setDirectory optionally set to true but purposely set it to false here. As you have promoted a stricter check on the segments, no validation here seems a bit off. Or am I missing something?

By default, setDirectory has `$validate=true`. The only reason I set it to `$validate=false` [when I call it in scanController](https://github.com/sneakyimp/CodeIgniter4/blob/bugs/issue-4294/system/Router/Router.php#L634) is because I have already [validated $segmentConvert a few lines before](https://github.com/sneakyimp/CodeIgniter4/blob/bugs/issue-4294/system/Router/Router.php#L624). My thinking was that it would be inefficient to validate the segment twice. There might be some better way to do this -- I considered having setDirectory return a boolean depending on whether the validation succeded or not, but setDirectory is a public method, and this would probably have introduced a Breaking Change.

sneakyimp commented:
I have no idea why these checks failed. They seem totally unrelated to my PR.

kenjis commented:
It is probably because ubuntu is now updated to 20.04.
But I don't know what's wrong.

paulbalandan commented:
Ubuntu latest is now 20.04 so only 7.4+ is built in in the image. We need to specify the extensions we need for lower php versions. I'll send a PR.

paulbalandan commented:
Please rebase,

sneakyimp reviewed (commented):
> Please rebase,

I have rebased and force-pushed. Please let me know if you have any additional thoughts.

kenjis commented on tests/system/Router/RouterTest.php:
```diff
@@ -575,4 +723,35 @@ public function testRegularExpressionPlaceholderWithUnicode()
+		$router->setDirectory('foo/bar/baz', false, true);
+		$router->handle('Some_controller/some_method/param1/param2/param3');
+
+		$this->assertEquals('foo/bar/baz/', $router->directory());
```
@sneakyimp Why is this test necessary?
Why does `setDirectory()` take precedence over URI?

kenjis replied:
I've found your answer.
https://github.com/codeigniter4/CodeIgniter4/pull/4307/files#r584375184
